plugins/modules/zos_find.py

```diff
@@ -0,0 +1,791 @@
+        "type": "CLUSTER"
+      },
+      {
+        "name": "SAMPLE.VSAM.DATA",
+        "type": "DATA"
+      }
+    ]
+matched:
+    description: The number of matched data sets found.
+    returned: success
+    type: int
+    sample: 49
+examined:
+    description: The number of data sets searched.
+    returned: success
+    type: int
+    sample: 158
+msg:
+    description: Failure message returned by the module.
+    returned: failure
+    type: str
+    sample: Error while gathering data set information
+stdout:
+    description: The stdout from a USS command or MVS command, if applicable.
+    returned: failure
+    type: str
+    sample: Searching dataset IMSTESTL.COMNUC
+stderr:
+    description: The stderr of a USS command or MVS command, if applicable.
+    returned: failure
+    type: str
+    sample: No such file or directory "/tmp/foo"
+rc:
+    description: The return code of a USS or MVS command, if applicable.
+    returned: failure
+    type: int
+    sample: 8
+"""
+
+import re
+import time
+import datetime
+import math
+
+from copy import deepcopy
+
+from ansible.module_utils.six import PY3
+from ansible.module_utils.basic import AnsibleModule
+
+from ansible_collections.ibm.ibm_zos_core.plugins.module_utils import (
+    vtoc, data_set, mvs_cmd
+)
+
+from ansible_collections.ibm.ibm_zos_core.plugins.module_utils.better_arg_parser import (
+    BetterArgParser
+)
+
+from ansible_collections.ibm.ibm_zos_core.plugins.module_utils.ansible_module import (
+    AnsibleModuleHelper
+)
+
+if PY3:
+    from shlex import quote
+else:
+    from pipes import quote
+
+
+def content_filter(module, patterns, content):
+    """ Find data sets that match any pattern in a list of patterns and
+    contains the given content
+
+    Arguments:
+        module {AnsibleModule} -- The Ansible module object being used in the module
+        patterns {list[str]} -- A list of data set patterns
+        content {str} -- The content string to search for within matched data sets
+
+    Returns:
+        dict[ps=set, pds=dict[str, str], searched=int] -- A dictionary containing
+        a set of matched "PS" data sets, a dictionary containing "PDS" data sets
+        and members corresponding to each PDS, an int representing number of total
+        data sets examined.
+    """
+    filtered_data_sets = dict(ps=set(), pds=dict(), searched=0)
+    for pattern in patterns:
+        rc, out, err = _dgrep_wrapper(pattern, content=content, verbose=True, ignore_case=True)
+        if rc > 4 and rc != 28:
+            module.fail_json(
+                msg="Non-zero return code received while executing ZOAU shell command 'dgrep'",
+                rc=rc, stdout=out, stderr=err
+            )
+        for line in err.splitlines():
+            if line and line.strip().startswith("BGYSC1005I"):
+                filtered_data_sets['searched'] += 1
+
+        for line in out.splitlines():
+            result = line.replace(content, "").split()
```

Comment:
Question: 

Correct me if I am misunderstanding the purpose of the code here: I think dgrep is going to return the file name and the matching line as a string, and you are attempting to get only the data set name by removing the value of `content` from the dgrep response.

Lets say you have a line in *MY.DATASET*
```
MY LINE HERE
```
And I specify I want to search for data sets with the string `LINE`, after you are done removing the value of `content` (i.e. `LINE`), would the characters `MY HERE` still be present in the resulting string?

Reply:
Hi @blakeinate, the intention of this piece of code is to determine whether each result returned by dgrep is a PS or a PDS/PDSE. Now, I know I can manually examine each entry and call LISTDS to check its DSORG. However, this is not going to scale very well. For results containing many entries, that would consume a lot of time.

Hence, I wanted to see if I can parse the string and check for any differences between results that are PS or PDS.
But yes, you are right. I have come up with a better checking scheme, which I will push soon.

Reply:
Makes sense to me, thanks for the clarification!